src/Tools/dotnet-monitor/Auth/ApiKeyAuthenticationOptionsChangeTokenSource.cs

```diff
@@ -0,0 +1,49 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+using System;
+using System.Threading;
+
+namespace Microsoft.Diagnostics.Tools.Monitor
+{
+    /// <summary>
+    /// Notifies that ApiKeyAuthenticationOptions changes when ApiAuthenticationOptions changes.
+    /// </summary>
+    internal sealed class ApiKeyAuthenticationOptionsChangeTokenSource :
+        IOptionsChangeTokenSource<ApiKeyAuthenticationOptions>,
+        IDisposable
+    {
+        private readonly IOptionsMonitor<ApiAuthenticationOptions> _optionsMonitor;
+        private readonly IDisposable _changeRegistration;
+
+        private ConfigurationReloadToken _reloadToken = new ConfigurationReloadToken();
+
+        public ApiKeyAuthenticationOptionsChangeTokenSource(
+            IOptionsMonitor<ApiAuthenticationOptions> optionsMonitor)
+        {
+            _optionsMonitor = optionsMonitor;
+            _changeRegistration = _optionsMonitor.OnChange(OnReload);
+        }
+
+        public string Name => Options.DefaultName;
+
+        public IChangeToken GetChangeToken()
+        {
+            return _reloadToken;
```

Comment:
It's been a while since i've looked at the details for this, but you have a read/write access to _reloadToken guarded by Interlocked and a read access on GetChangeToken with no memory barriers. IIRC this can cause stale data on some threads if multiple threads are attempting to call GetChangeToken.

Reply:
It seems that the entire execution is single threaded, so when the change token raises the change callback, all of the handlers can back in one right the other to get the new change token.